### DISCUSSION TITLE
cycle-free graph violated - partition depends on itself

### DISCUSSION THREAD

**Discussion by lumawu** [Xilinx/finn#431] [Q&A] [answered]
Hey there, I am currently trying to go through the end2end_cnv toolflow with a network I modified on my own using brevitas.
However, during partitioning I run into the following error while creating a dataflow partition:
![error](https://user-images.githubusercontent.com/72506237/141991692-f443c962-48b5-4a31-829f-299885b3947d.png)

I used the example network on the pytorch tutorials as basis:
![examplenet](https://user-images.githubusercontent.com/72506237/141991741-7cbec717-d50b-4c8d-ab22-7c68fdbb96fd.png)

Here are the modifications I made to it, training was done with a bit width of 1 for weights and activation and an input width of 8:
![exampleNetModified](https://user-images.githubusercontent.com/72506237/141991795-1e403557-9866-43c8-ac8f-5378280cc574.png)

the onnx model has the following topology right before the partitioning call:
![netBeforePartitioning](https://user-images.githubusercontent.com/72506237/141991842-8cf53860-d1ee-4569-a215-f3bf710c161d.png)

Brevitas is on v0.6, Finn on v0.7, pytorch on v1.7.1
This is my first time using my own network and I couldn't find much info on the error on the net, so I hope I can find help here

**Comment by fpjentzsch** [comment, marked as answer]
Hi, the problem is that the MaxPool layer was not correctly streamlined and converted to an HLS node ("StreamingMaxPool_Batch").

Do you apply the `MakeMaxPoolNHWC()` and `InferStreamingMaxPool()` transformations during streamlining and HLS-conversion, respectively?

**Reply by lumawu** [reply]
Thanks for the fast reply, yes both transformations have been applied. The order in which all transformations are called are the same as in the end2end_conv example notebook.

Here are the respective parts:

from finn.transformation.streamline import Streamline
from finn.transformation.lower_convs_to_matmul import LowerConvsToMatMul
from finn.transformation.bipolar_to_xnor import ConvertBipolarMatMulToXnorPopcount
import finn.transformation.streamline.absorb as absorb
from finn.transformation.streamline.reorder import MakeMaxPoolNHWC, MoveScalarLinearPastInvariants
from finn.transformation.infer_data_layouts import InferDataLayouts
from finn.transformation.general import RemoveUnusedTensors

model = ModelWrapper(build_dir + "/end2end_cnv_w1a1_pre_post.onnx")
model = model.transform(MoveScalarLinearPastInvariants())
model = model.transform(Streamline())
model = model.transform(LowerConvsToMatMul())
model = model.transform(MakeMaxPoolNHWC())
model = model.transform(absorb.AbsorbTransposeIntoMultiThreshold())
model = model.transform(ConvertBipolarMatMulToXnorPopcount())
model = model.transform(Streamline())
model = model.transform(absorb.AbsorbScalarMulAddIntoTopK())
model = model.transform(InferDataLayouts())
model = model.transform(RemoveUnusedTensors())
model.save(build_dir + "/end2end_cnv_w1a1_streamlined.onnx")

import finn.transformation.fpgadataflow.convert_to_hls_layers as to_hls
from finn.transformation.fpgadataflow.create_dataflow_partition import (
    CreateDataflowPartition,
)
from finn.transformation.move_reshape import RemoveCNVtoFCFlatten
from finn.custom_op.registry import getCustomOp
from finn.transformation.infer_data_layouts import InferDataLayouts

mem_mode = "decoupled"

model = ModelWrapper(build_dir + "/end2end_cnv_w1a1_streamlined.onnx")
model = model.transform(to_hls.InferBinaryStreamingFCLayer(mem_mode))
model = model.transform(to_hls.InferQuantizedStreamingFCLayer(mem_mode))
model = model.transform(to_hls.InferLabelSelectLayer())
model = model.transform(to_hls.InferThresholdingLayer())
model = model.transform(to_hls.InferConvInpGen())
model = model.transform(to_hls.InferStreamingMaxPool())
model = model.transform(RemoveCNVtoFCFlatten())
model = model.transform(absorb.AbsorbConsecutiveTransposes())
model = model.transform(InferDataLayouts())
model.save(build_dir + "/end2end_cnv_w1a1_intermediaryhls.onnx")
showInNetron(build_dir+"/end2end_cnv_w1a1_intermediaryhls.onnx")
parent_model = model.transform(CreateDataflowPartition())
parent_model.save(build_dir + "/end2end_cnv_w1a1_dataflow_parent.onnx")
sdp_node = parent_model.get_nodes_by_op_type("StreamingDataflowPartition")[0]
sdp_node = getCustomOp(sdp_node)
dataflow_model_filename = sdp_node.get_nodeattr("model")
dataflow_model = ModelWrapper(dataflow_model_filename)
dataflow_model.save(build_dir + "/end2end_cnv_w1a1_dataflow_model.onnx")

**Reply by fpjentzsch** [reply]
Could you show me the .onnx model before `MakeMaxPoolNHWC()` is applied?

**Reply by lumawu** [reply]
yes of course, here it is:
![beforeMaxPoolNHWC](https://user-images.githubusercontent.com/72506237/142021734-c268e234-7b1b-4406-9711-e6b47034d145.png)

**Reply by fpjentzsch** [reply]
I see, you can try to put a second `MakeMaxPoolNHWC()`, followed by an `AbsorbConsecutiveTransposes()`, behind the `AbsorbTransposeIntoMultiThreshold()`, similar to how we do it in the default "build_dataflow" build steps:
https://github.com/Xilinx/finn/blob/dev/src/finn/builder/build_dataflow_steps.py#L260

**Reply by lumawu** [reply]
yes this solved the problem, thank you!

**Reply by roderick1014** [reply]
Hi! I also occur this problem when I trying to implement my custom network. My jupyter notebook code is same as the @lumawu , here is my .onnx model. I am sure that the  the MaxPool layer has correctly streamlined.
![YOLOv1_Quantized_3_streamlined onnx](https://user-images.githubusercontent.com/73574008/147794498-c18036ac-b54f-4790-89b8-879553213444.png)

**Reply by fpjentzsch** [reply]
Hi @roderick1014, I think the problem might be the double flatten node. To make sure, you can look at the .onnx after streamlining and after HLS conversion and see where it breaks into 2 partitions. Do you have any idea why there are 2 flatten nodes instead of one?

**Reply by caoimhelynch** [reply]
Hi there @fpjentzsch, I hope you are still open to answering another version of this same issue. I am getting the error 'AssertionError: cycle-free graph violated: partition depends on itself', when creating the dataflow partition 
Here is my model .onnx model before MakeMaxPoolNHWC()
![image](https://user-images.githubusercontent.com/59800744/154995479-e0a2786b-d184-4c10-ad57-3229d1ac7280.png)
![image](https://user-images.githubusercontent.com/59800744/154995555-109b42e4-92b9-4d9e-961f-d6d0a27e1bbd.png)
![image](https://user-images.githubusercontent.com/59800744/154995605-9728ad1a-be1d-484b-92d0-73f09712bbec.png)
I attempted to put a second MakeMaxPoolNHWC() followed by AbsorbConsecutiveTransposes() but it did not work for my issue. Appreciate any help thanks!

**Reply by fpjentzsch** [reply]
What looks suspicious to me:

- You apply MaxPool to odd feature map dimensions (e.g. (41, 41) -> (20, 20)). As you see [here](https://github.com/Xilinx/finn/blob/fd45721c86f84d634ef23ccec65aa1811f38baa9/src/finn/transformation/fpgadataflow/convert_to_hls_layers.py#L355), kernel size must divide the dimensions evenly for "StreamingMaxPool_Batch". Maybe you can apply some padding to solve this?
- There is a MultiThreshold ->MaxPool -> MultiThreshold sequence, which looks redundant. Maybe this causes problems during streamlining.

To narrow down which layers cause trouble, you can look at the .onnx model after the streamlining and convert-to-hls steps.

**Reply by caoimhelynch** [reply]
Hi thanks for getting back to me. This is my model just before the createDataflowPartition() is called. 
![image](https://user-images.githubusercontent.com/59800744/155126271-55b5250a-72ec-402e-ad6b-a78c9b8e1f1b.png)
![image](https://user-images.githubusercontent.com/59800744/155126350-0fbf78aa-f7e1-4114-9b65-4792d7a16a7c.png)
![image](https://user-images.githubusercontent.com/59800744/155126401-83d567c7-8055-434d-92e6-e5a2d3e2746b.png)
I think this shows the same initial issue of this discussion where the maxpool has not correctly streamlined. Will attempt to alter the dimensions as suggested. Am I correct in thinking that maxpool not streamlining is the issue?

**Reply by fpjentzsch** [reply]
Yes, because there is at least one "MaxPool" node that was not converted to a "StreamingMaxPool_Batch".

**Reply by caoimhelynch** [reply]
@fpjentzsch I attempted your suggestion of adding padding but it didn't seem to help the issue and started causing problems in other areas. I also don't think it is the maxPoolLayer with 41x41 that is causing the issue, it is the 5th maxpool layer in the model. 

From debugging in the MakeMaxPoolNHWC() method, it seems that the consumer.op_type and producer.op_type are returning as 'MultiThreshold' rather than 'Transpose' which is why that layer does not get streamlined. Are you aware of any reason why this would happen? 

Appreciate any help!

**Reply by fpjentzsch** [reply]
In that case the problem seems to be the MaxPool at the transition between the convolutional and fully-connected parts of the model. It seems like the current sequence of transformation steps expects MaxPool nodes to be followed by yet another convolution layer, which should have the form (Transpose->Im2Col->MatMul->Transpose->MultiThreshold (optional)) after lowering.

Looking at your CNV to FC transition (...->Transpose->**MultiThreshold**->MaxPool->MultiThreshold->Flatten->...), maybe it will still work with a MaxPool node after the last convolution layer if you remove the **first Multithreshold**. As you have pointed out, `MakeMaxPoolNHWC()` needs a Transpose either before or after the MaxPool. However, the MaxPool will then operate on the un-quantized output of MatMul...

You could also try to call `MakeMaxPoolNHWC()` again after inferring the Thresholding_Batch nodes etc. This should convert the (Transpose->MaxPool->Transpose) that is left in your graph to (Transpose->Transpose->MaxPoolNHWC) and the consecutive Transpose nodes could be absorbed.

**Reply by caoimhelynch** [reply]
Really appreciate you getting back to me. 
I do not fully understand what you are saying. Are you saying that I need to add a convolutional layer after the maxpool(or are you just saying that seems to be expected)? Or that I should remove the multithreshold before the maxpool layer? 
If that is the case how would I go about removing this multithreshold layer? I am not entirely sure on that.

**Reply by fpjentzsch** [reply]
I just updated my previous message. I suggest you try calling `MakeMaxPoolNHWC()` again (on the model state shown in your previous screenshot), followed by `AbsorbConsecutiveTransposes()` and `InferStreamingMaxPool()`.

If you are willing to alter your model, you could remove the final MaxPool or add a convolution layer, so that there is no final MaxPool before the fully-connected layers begin.

**Reply by caoimhelynch** [reply]
Okay so I tried calling MakeMaxPoolNHWC() then  AbsorbConsecutiveTransposes() along with with InferStreamingMaxPool() so that all of the above are called more than once. It seems to have successfully steamlined the maxpool layer however I am still getting the exact same error.
Here is the streamlined model right before the line CreateDataflowPartition()
![image](https://user-images.githubusercontent.com/59800744/155557718-a8a561c9-2252-4705-ad22-4c083cb3d80d.png)
![image](https://user-images.githubusercontent.com/59800744/155557815-955b1273-1e93-46d9-af84-f4d12b803f72.png)
![image](https://user-images.githubusercontent.com/59800744/155557873-20f08760-4006-4b1c-8004-d6988c4e29d9.png)
I was hoping the unstreamlined maxpool layer was the source of the error but now that this is fixed and I am still getting the same error I am confused. Is there anything else that could be causing this error to happen? 
Sorry for so many questions I am new to finn and new to creating models in general so really appreciate the help you are providing!

**Reply by fpjentzsch** [reply]
The final nodes (Mul, Add, TopK) are not converted to HLS nodes yet. Try `InferChannelwiseLinearLayer()` for the Mul/Add  and `InferLabelSelectLayer()` for the TopK, followed by `GiveUniqueNodeNames()` to clean up the model again.

**Reply by fpjentzsch** [reply]
Also, the MaxPool dimension problem persists for the 41x41 and 13x13 case, as your nodes were streamlined to MaxPoolNHWC, but not converted to StreamingMaxPool_Batch.

**Reply by caoimhelynch** [reply]
Ah okay wasn't sure what the mul add and Topk should look like currently attempting to convert those layers.

To fix this 41x41 and 13x13, is there a way to add padding that doesn't require adding changing the model? Trying to change that caused a lot of issues for me both in initially exporting to onnx and when the onnx model was being converted to HLS.

**Comment by KeciaHH** [comment]
Hi @fpjentzsch, hope you are still answering this issue. I have met a similar problem, this is what my model looks like:
![WechatIMG26942](https://user-images.githubusercontent.com/49864650/229440294-2a359e86-5eeb-42e8-9113-d2503ef496de.png)
I used the default build_dataflow_steps script in the src folder and commented second Streamline() in the step_streamline() function (cuz it may cause some errors about InferShape).

**Reply by fpjentzsch** [reply]
Hi @KokyoK,

unfortunately the default build_dataflow steps will not work for a model with residual connections. Maybe you can take inspiration from the ResNet-50 finn-example on how to adjust the streamlining and convert_to_hls steps accordingly. You can find the code here:

https://github.com/Xilinx/finn-examples/blob/main/build/resnet50/build.py
https://github.com/Xilinx/finn-examples/blob/main/build/resnet50/custom_steps.py

**Reply by pkdeep** [reply]
Hi @fpjentzsch ,
 Can you help with this. I am also trying a network with residual connections, and facing the same issue. Few layer have not been converted to hls layer, hence getting cycle-free graph violated error.
![partial_converted_parallel_model](https://github.com/Xilinx/finn/assets/78947824/6100cc26-88bf-4dbf-85c1-2b066317b9de)

Thanks

**Reply by fpjentzsch** [reply]
Hi, unfortunately I currently do not have the time to dive into your specific model. Maybe this recent thread can help you? https://github.com/Xilinx/finn-examples/issues/67

**Comment by sansi-zhang** [comment]
Hi @fpjentzsch, 
I ran into the same problem and I tried the same thing with ResNet-50 but it still failed.

### In detail
I ran into a problem while executing ```model.transform(CreateDataflowPartition())```.

![image](https://github.com/Xilinx/finn/assets/85163357/2eebb517-6cbb-472c-b6ca-659e38efe57d)

I made some adjustments and tests (see https://github.com/Xilinx/finn/issues/860#issuecomment-1649853296 for details), but the error still occurred.

Then I looked at the answers in this issue, and I used the ResNet-50 process to re-tune my code, but I was still stuck.

Part of the onnx file when resnet-50 is not used, see  https://github.com/Xilinx/finn/issues/860#issuecomment-1649853296

Part of the onnx file when using resnet-50, as follows.

![image](https://github.com/Xilinx/finn/assets/85163357/199d1708-3817-42b6-af2b-ad1fddbe74a1)

![image](https://github.com/Xilinx/finn/assets/85163357/4b9de4be-c508-405d-999c-21051bcb0ba2)

I guessed that it was due to the concat operation, so I stopped using concat and split all the output, but still had the same problem.
![image](https://github.com/Xilinx/finn/assets/85163357/32a4bafd-4e47-47f9-af11-d022ebef3a2d)

### I hope you can help me solve this problem.
## Thanks again.

**Reply by pkdeep** [reply]
Did anyone get through this problem successfully? I am stuck with the similar problem on a fully connected layer. Even tfc_2w_2a is not working for me in bnn_pynq example notebook.
